### ISSUE TITLE
How to integrate tflite models in flutter

### ISSUE THREAD

**Issue by AkshatRastogi-1nC0re** [lshug/first-order-model-tf#16]
I am new to tensorflow, I wanted to know all possible ways in which I could integrate this into android.
I know that because of the heavy model, it cannot be deployed in heroku to use as an API. Can you please guide us on how that can be achieved?

**Comment by lshug** [comment]
Hello. The functioning of this model is pretty complex. It has three sub-modules, each with either multiple inputs/outputs. and they need to be used in a specific order. Automatic tflite integration tools won't be of much help.

For flutter integration, you'll want to use [tflite_flutter](https://pub.dev/packages/tflite_flutter). Probably the best way to do so will be to write a wrapper function for the inference call of each of the three modules (kp_detector, process_kp_driving, generator), which will create input and output buffers, put them in maps according to the mappings shown in json files generated alongside tflite files on python side, runs interpreter.runForMultipleInputs, and then returns the values of the output buffers. After you have wrappers for all three modules, you'll have to implement the inference schema:

* Run kp_detector on source image, generating kps and jacobians for the source image
* Run kp_detector on the first frame of the driving video, generating kps and jacobians for the initial driving video frame
* For each frame/batch of frames of the driving video:
  * Run kp_detector, generating driving video frame kps and jacobians
  * Run process_kp_driving to generate normalized kps and jacobians using kps and jacobians of the source image, current frame, and initial frame.
  * Finally, run generator on source image, normalized kps and jacobians, and source kps and jacobians to generate the predicted frame/frames. Append them to a list.
* Once the process is done on all frames, concatenate all predicted frames, and return as a video.

Unfortunately I can't help you in any more detail. Best of luck.

**Comment by AkshatRastogi-1nC0re** [comment]
Any idea, how much time will it take to run this natively on mobile. On PC it is taking approximately 3-4 minutes with tflite models and a 1660ti GPU ?

**Comment by lshug** [comment]
On python side, tflite isn't GPU accelerated, so the time it takes is a good indicator of how well the model performs on CPU.

Which is to say, without further hardware acceleration through the use of delegates, the models are going to be slow. Even on current flagships and with builds with optimized architecture (predictiononly, staticbatchsize, prescale), it will take about ~6-8 seconds per single frame (haven't tried >1 batch sizes on mobile, but RAM is a limitation so you realistically can't go much higher than 4). The good news is that almost all of this time is spent on performing fused convolution-batchnorm ops, so GPU acceleration should provide significant speedup. However, currently the models are not compatible with the GPU delegate, due to per-op incompatibility (the two big hurdles are float-to-int casting and tfl::GatherOp).
